cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -229,6 +229,8 @@ public TwillController launch(TwillApplication twillApplication, Iterable<String
             // TokenSecureStoreUpdater.update() ignores parameters
             twillPreparer.addSecureStore(secureStoreUpdater.update(null, null));
           }
+
+          System.setProperty("user.name", cConf.get(Constants.CFG_HDFS_USER));
```

Comment:
Also, in develop branch, it's better to copy the FileContextLocation from twill and change the implementation of `getHomeLocation` to shadow the one in twill.